lib/crypto/c_src/pbkdf2_hmac.c

```diff
@@ -0,0 +1,99 @@
+/*
+ * %CopyrightBegin%
+ *
+ * Copyright Ericsson AB 2010-2021. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * %CopyrightEnd%
+
+ */
+
+#include "common.h"
+#include "pbkdf2_hmac.h"
+#include "digest.h"
+
+ERL_NIF_TERM pbkdf2_hmac_nif(ErlNifEnv* env, int argc,
+                             const ERL_NIF_TERM argv[])
+{
+    ErlNifBinary pass, salt, out;
+    ErlNifUInt64 iter, keylen;
+    struct digest_type_t* digp = NULL;
+    const EVP_MD* digest;
+
+    ASSERT(argc == 5);
+
+    if ((digp = get_digest_type(argv[0])) == NULL)
+        goto bad_arg;
+    if (digp->md.p == NULL)
+        goto bad_arg;
+
+    switch (EVP_MD_type(digp->md.p))
```

Comment:
This `switch .... { ... }` is a bit too complicated.  The `digest = EVP_sha1();` etc are not needed. The EVP_MD* is already present in `digp->md.p`
The check of that the EVP_MD is sha1 or sha2 could be done by adding a new flag in digest.h, add it to the definitions of digests in digest.c and finally as a test in your code that it is present. See for instance cipher.c, cipher.h.
If it seem too complicated I can add do it later.

Reply:
@HansN I've done something don't hesitate if it's not why you have it in mind.

Reply:
Your change is exactly what I tried to describe!